### PR TITLE
feat(Workspaces): improve invitation management

### PR DESCRIPTION
**Pull request by cheikhgwane**
Improve workspace invitation management.

## Changes

Please list / describe the changes in the codebase for the reviewer(s).

- Allow admin to delete pending invitations.
- Re-use existing invitation (Pending) if admin want to invite the same user.

## How/what to test

A few words about what needs to be tested, along with specific testing instructions if needed.

## Screenshots / screencast

If relevant, please add some screenshots or a short video.

### CODE REVIEW

File: hexa/workspaces/schema/mutations.py [hunk 1]
```diff
@@ -150,13 +150,19 @@ def resolve_invite_workspace_member(_, info, **kwargs):
...
             }
         except User.DoesNotExist:
-            # If the user does not exist, we create an invitation to create an account and join the workspace
-            invitation = WorkspaceInvitation.objects.create_if_has_perm(
-                principal=request.user,
-                workspace=workspace,
-                email=input["userEmail"],
-                role=input["role"],
-            )
+            # If the user does not exist we first
```
Comment (qgerome) [inline hexa/workspaces/schema/mutations.py:153]: Better to let the code crash thanks to a unique constraint

**Review by qgerome — CHANGES_REQUESTED** [review]
From what I see in the task, we should also have a "resend invitation" mutation.

File: hexa/workspaces/schema/mutations.py [hunk 2]
```diff
@@ -196,7 +195,7 @@ def resolve_invite_workspace_member(_, info, **kwargs):
 
 
 @workspace_mutations.field("joinWorkspace")
-def resolver_join_workspace(_, info, **kwargs):
+def resolve_join_workspace(_, info, **kwargs):
```
Comment (qgerome) [inline hexa/workspaces/schema/mutations.py:198]: thanks

File: hexa/workspaces/schema/mutations.py [hunk 3]
```diff
@@ -266,8 +265,57 @@ def resolver_join_workspace(_, info, **kwargs):
...
+def resolve_resend_workspace_invitation(_, info, **kwargs):
+    request: HttpRequest = info.context["request"]
+    input = kwargs["input"]
+
+    try:
+        invitation = WorkspaceInvitation.objects.get(id=input["invitationId"])
+        if not request.user.has_perm("workspaces.manage_members", invitation.workspace):
+            raise PermissionDenied
+
+        if invitation.status == WorkspaceInvitationStatus.ACCEPTED:
```
Comment (qgerome) [inline hexa/workspaces/schema/mutations.py:278]: I think you can use the same use case than the invitation not found above

File: hexa/workspaces/schema/mutations.py [hunk 4]
```diff
@@ -266,8 +265,57 @@ def resolver_join_workspace(_, info, **kwargs):
         }
 
 
+@workspace_mutations.field("resendWorkspaceInvitation")
+def resolve_resend_workspace_invitation(_, info, **kwargs):
+    request: HttpRequest = info.context["request"]
+    input = kwargs["input"]
+
+    try:
+        invitation = WorkspaceInvitation.objects.get(id=input["invitationId"])
```
Comment (qgerome) [inline hexa/workspaces/schema/mutations.py:274]:
```suggestion
        invitation = WorkspaceInvitation.objects.filter(status=WorkspaceInvitationStatus.PENDING).get(id=input["invitationId"])
```

File: hexa/workspaces/schema/mutations.py [hunk 5]
```diff
@@ -266,8 +265,57 @@ def resolver_join_workspace(_, info, **kwargs):
...
+        if not request.user.has_perm("workspaces.manage_members", invitation.workspace):
+            raise PermissionDenied
+
+        if invitation.status == WorkspaceInvitationStatus.ACCEPTED:
+            raise AlreadyExists("Invitation already accepted.")
+
+        token = invitation.generate_invitation_token()
+        send_mail(
```
Comment (qgerome) [inline hexa/workspaces/schema/mutations.py:282]: If it's 99% the same code as the initial invite, it's better to make a function of this and call it at the different places.

File: hexa/workspaces/schema/mutations.py [hunk 6]
```diff
@@ -266,8 +265,57 @@ def resolver_join_workspace(_, info, **kwargs):
...
+        return {
+            "success": False,
+            "errors": ["PERMISSION_DENIED"],
+        }
+    except WorkspaceInvitation.DoesNotExist:
+        return {
+            "success": False,
+            "errors": ["INVITATION_NOT_FOUND"],
+        }
+    except AlreadyExists:
```
Comment (qgerome) [inline hexa/workspaces/schema/mutations.py:310]: This can be can be removed then

**Review by qgerome — CHANGES_REQUESTED** [review]

**Review by qgerome — APPROVED** [review]

**Review by qgerome — CHANGES_REQUESTED** [review]
In fact, I don't see the constraint on the workspace and email address that would be useful to not having multiple invitations with the same email addresses

**Review by qgerome — CHANGES_REQUESTED** [review]
You still need to catch the error thrown by Postgresql and return a proper error to the frontend (AlreadyExists)

**Comment by cheikhgwane** [comment]
> You still need to catch the error thrown by Postgresql and return a proper error to the frontend (AlreadyExists)

I've updated the resolver inviteWorkspaceMember. For the resend we do no try to create a new one.

**Review by qgerome — APPROVED** [review]

**Pull request merged**